cldf/parameters.csv

```diff
@@ -1,96 +1,96 @@
 ID,Name,Concepticon_ID,Concepticon_Gloss
-all,All,98,ALL
-ash,Ashes,646,ASH
-bark,Bark (tree),1204,BARK
-belly,Belly,1251,BELLY
-big,Big,1202,BIG
-bird,Bird,937,BIRD
-bite,Bite (v.),1403,BITE
-black,Black,163,BLACK
-blood,Blood,946,BLOOD
-bone,Bone,1394,BONE
-breast,Breast,1402,BREAST
-burnsomething,Burn (tr. v.),141,BURN (SOMETHING)
-cloud,Cloud,1489,CLOUD
-cold,Cold (air),1287,COLD
-come,Come,1446,COME
-die,Die,1494,DIE
-dog,Dog,2009,DOG
-drink,Drink (v.),1401,DRINK
-dry,Dry (adj.),1398,DRY
-ear,Ear,1247,EAR
-eat,Eat,1336,EAT
-egg,Egg,744,EGG
-eye,Eye,1248,EYE
-fatorganicsubstance,Fat (n.),323,FAT (ORGANIC SUBSTANCE)
-fire,Fire,221,FIRE
-fish,Fish (n.),227,FISH
-flyinsect,Fly (n.),1504,FLY (INSECT)
-foot,Foot,1301,FOOT
-give,Give,1447,GIVE
-go,Go,695,GO
-good,Good,1035,GOOD
-grass,Grass,606,GRASS
-hairhead,Hair (head),2648,HAIR (HEAD)
-hand,Hand,1277,HAND
-head,Head,1256,HEAD
-hear,Hear,1408,HEAR
-heart,Heart,1223,HEART
-hornanatomy,Horn,1393,HORN (ANATOMY)
-i,I,1209,I
-kill,Kill,1417,KILL
-knowsomething,Know,1410,KNOW (SOMETHING)
-knee,Knee,1371,KNEE
-leaf,Leaf,628,LEAF
-liver,Liver,1224,LIVER
-long,Long,1203,LONG
-louse,Louse,1392,LOUSE
-man,Man,1554,MAN
-many,Many,1198,MANY
-meat,Meat,634,MEAT
-moon,Moon,1313,MOON
-mountain,Mountain,639,MOUNTAIN
-mouth,Mouth,674,MOUTH
-name,Name,1405,NAME
-neck,Neck,1333,NECK
-new,New,1231,NEW
-night,Night,1233,NIGHT
-nose,Nose,1221,NOSE
-one,One,1493,ONE
-rainprecipitation,Rain (n.),658,RAIN (PRECIPITATION)
-red,Red,156,RED
-road,Road,667,ROAD
-root,Root,670,ROOT
-sand,Sand,671,SAND
-say,Say,1458,SAY
-see,See,1409,SEE
-seed,Seed,714,SEED
-sit,Sit,1416,SIT
-skinhuman,Skin (human),2613,SKIN (HUMAN)
-sleep,Sleep (v.),1585,SLEEP
-small,Small,1246,SMALL
-smokeexhaust,Smoke (n.),778,SMOKE (EXHAUST)
-snake,Snake,730,SNAKE
-stand,Stand,1442,STAND
-star,Star,1430,STAR
-stone,Stone,857,STONE
-sun,Sun,1343,SUN
-swim,Swim,1439,SWIM
-tail,Tail,1220,TAIL
-thin,Thin,2308,THIN
-this,This,1214,THIS
-three,Three,492,THREE
-thou,Thou,1215,THOU
-tongue,Tongue,1205,TONGUE
-tooth,Tooth,1380,TOOTH
-tree,Tree,906,TREE
-two,Two,1498,TWO
-warm,Warm,1232,WARM
-water,Water,948,WATER
-we,We,1212,WE
-wet,Wet,1726,WET
-what,What?,1236,WHAT
-white,White,1335,WHITE
-who,Who?,1235,WHO
-woman,Woman,962,WOMAN
-you,Ye,1213,YOU
+Kitchen-2009-95-1,All,98,ALL
+Kitchen-2009-95-2,Ashes,646,ASH
+Kitchen-2009-95-3,Bark (tree),1204,BARK
```

Comment:
In language IDs, I now turn to have very explicit hand-crafted IDS, that express the language name, but in ascified form, easy to use in nexus, and shortened, where needed.

For concepts, this is difficult, as we have concept IDs from concepticon, which are lengthy, and we cannot use concepticon_ID as an ID, as they may well be missing. So we need a way to have consistent IDs anyway. 

One way would be to slug the concept-glosses in the data, which is, however, dangerous, as there's no check for mergers. Another way is to use the id_factory function, e.g., to use a combination of a number and a concept-gloss, which would guarantee uniqueness. 

But then there's issue [130](https://github.com/lexibank/pylexibank/issues/130), where we discuss of how to facilitate all of this. 

And you can add concepts in two ways: from concepticon, and from concepts.tsv. I think I'd opt for a more rigorous ID-factory, that would, e.g., use the last part of a concepticon-id, and then combine this with the concepticon_ID (which can also be empty):

```python
Parameter_ID = lambda x: x['ID'].split('-')[-1]+'-'+slug(x['Concepticon_ID'])
```
This would also have an advantage for the manual checking of the form-file, and problems resulting from mixed concept mappings would show up earlier.

Reply:
I'd be pragmatic here: Whenever you have a better **unique**, **ascii-only** column - possibly the gloss - use it. But as soon as you'd have to sluggify or similar, fall back to the conceptlist IDs.

Reply:
I guess, that's fine. We should see how to solve that issue I mentioned, however, as this would also allow us to make it clearer that IDs are not created on the fly (so I'd like to get rid of the id_factory-lambda construct, which is potentially dangerous here).

Reply:
I fully agree here -- it should be as close as possible to the raw data, and language identifiers should be NEXUS-friendly (in the sense, at least, of SplitsTree and BEAST2 friendly).

I have that function of mine that wraps `slug()` and, when fed a list of all the labels, makes sure to return a dictionary with unique slugged keys, adding material only when needed (I sent it to @LinguList and @xrotwang some months ago, as an half-baked idea). I could make a PR, if you are interested, but using it would mean changing the logic in the code of most datasets.

Reply:
My opinion here is very clear now, after having refined identifiers for some 20+ datasets: The identifiers for languages should ideally be handmade, and slugging functions can be used, of course, in order to create them, but it is always better, when working on collections where one is seriously interested in, if an expert gives good names. E.g., in our case, I just sat down and fixed names in smaller datasets of 30+ languages (I would not do that for ABVD now). But the identifiers should definitely be listed in `etc/languages.csv` and not be created on the fly.

Reply:
Agreed that it is better to have the identifiers in `languages.csv` and in the concept lists.

However, and despite being more interested in readable parameter identifiers than language ones, having a semi-automated slugging from the raw names helps more in debugging than in post-cldf tasks. My philosophical restriction to language names modified by us (or experts working with us) is that it is not much different from non-explicit data modification (especially if the new identifier end up being used elsewhere, like in an analysis), and besides using the language id (which only needs to be a unique string) for what the Glottolog mapping is intended (i.e., linking to the right variety).

If the raw data says the language name is "Baba", I would prefer to have an identifier `dataset-lang-baba` and not some `dataset-lang-sisi`, even in those cases where we are sure the source misidentified it (and I won't venture about alternative names and all the political issues it might bring up).

Reply:
Just to leave it here as an idea for a command-line tool that could be used for language and concept identifiers (see [gist](https://gist.github.com/tresoldi/2f73355330b88e7068c9590b96f42ea9)). Running `./slugids.py languages.csv` on this repo would return:

```
ID,Name,Glottocode,Sources
akkadian,Akkadian,akka1240,Rabin1975;Kitchen2009
amharic,Amharic,amha1245,Bender1971;Kitchen2009
aramaic,Aramaic,olda1245,Rabin1975;Kitchen2009
argobba,Argobba,argo1244,Bender1971;Kitchen2009
chaha,Chaha,chah1248,Bender1971;Kitchen2009
gafat,Gafat,gafa1240,Bender1971;Kitchen2009
geez,Gɛ'ɛz,geez1241,Bender1971;Kitchen2009
geto,Geto,gyet1240,Bender1971;Kitchen2009
gibbali,Gibbali,sheh1240,Kitchen2009
harari,Harari,hara1271,Bender1971;Kitchen2009
harsusi,Harsusi,hars1241,Kitchen2009
hebrew,Hebrew,hebr1245,Rabin1975;Kitchen2009
innemor,Innemor,inor1238,Bender1971;Kitchen2009
mehri,Mɛhri,mehr1241,Kitchen2009
mesmes,Mesmes,mesm1243,Bender1971;Kitchen2009
mesqan,Mesqan,mesq1240,Bender1971;Kitchen2009
moroccanarabic,Moroccan Arabic,moro1292,Sobelman1963;Kitchen2009
ogadenarabic,Ogadɛn Arabic,ogad1238,Bender1971;Kitchen2009
soddo,Soddo,sodd1242,Bender1971;Kitchen2009
soqotri,Soqotri,soqo1240,Leslau1938;Kitchen2009
tigre,Tigrɛ,tigr1270,Bender1971;Kitchen2009
tigrinya,Tigrinya,tigr1271,Bender1971;Kitchen2009
ugaritic,Ugaritic,ugar1238,Rabin1975;Kitchen2009
walani,ʷalani,wola1253,Bender1971;Kitchen2009
zway,Zway,zayy1238,Bender1971;Kitchen2009
```
Which could be corrected/changed by the maintainer in cases such as misidentified languages. The script takes care of duplicates, so that, if we had a second language whose id would be `zway`, the first would be corrected to `zway-a` and the second to `zway-b` (particularly useful for concepts).

Reply:
I'd recommend to propose this via a PR as a commandline tool of pylexibank (or cldfbench). This could be part of the workflow and indeed save some time.

Reply:
This is already merged, right? We need to work towards making py2 now, so all enhancements that somebody comes up with and which seem useful can already go into it. At least I have a couple of things as well which I'll propose (templates, etc.). So feel free to prepare a PR, @tresoldi.

Reply:
Can I get a final verdict on this PR here concerning the IDs?

Reply:
Then I'm confused - isn't this the way it is currently?

```
ID,Name,Concepticon_ID,Concepticon_Gloss
Kitchen-2009-95-1,All,98,ALL
Kitchen-2009-95-2,Ashes,646,ASH
Kitchen-2009-95-3,Bark (tree),1204,BARK
```